### PR TITLE
[Enhancement] Add force rule based mv rewrite

### PR DESCRIPTION
**Pull request by ABingHuang**
Why I'm doing:
the best rewritten result can not be selected because incorrect cost estimate of plan.
What I'm doing:
Add a force rule based strategy for mv rewrite. Use session variable 'eable_force_rule_based_mv_rewrite' to control it. And the default is false.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/pattern/Pattern.java [hunk 1]
```diff
@@ -141,6 +141,10 @@ public boolean matchWithoutChild(OptExpression expression) {
             return true;
         }
 
+        if (isPatternMultiJoin() && isMultiJoin(expression.getOp().getOpType())) {
+            return true;
+        }
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/pattern/Pattern.java:146]: Why add this?

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 2]
```diff
@@ -1445,6 +1447,9 @@ public String getConnectorSinkCompressionCodec() {
     @VarAttr(name = ENABLE_MATERIALIZED_VIEW_REWRITE_PARTITION_COMPENSATE, flag = VariableMgr.INVISIBLE)
     private boolean enableMaterializedViewRewritePartitionCompensate = true;
 
+    @VarAttr(name = ENABLE_FORCE_RULE_BASED_MV_REWRITE)
+    private boolean enableForceRuleBasedMvRewrite = false;
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:1451]: Can this be true by default?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/SingleTableMvRewriteRule.java [hunk 3]
```diff
@@ -1,76 +0,0 @@
...
-import com.starrocks.sql.optimizer.rule.Rule;
-import com.starrocks.sql.optimizer.rule.RuleSetType;
-import com.starrocks.sql.optimizer.rule.RuleType;
-
-import java.util.List;
-
-// used to rewrite query with single table scan pattern to mv
-public class SingleTableMvRewriteRule extends Rule {
-    public SingleTableMvRewriteRule() {
-        super(RuleType.TF_MV_CBO_SINGLE_TABLE_REWRITE_RULE, Pattern.create(OperatorType.PATTERN));
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/SingleTableMvRewriteRule.java:34]: Why remove this? This is not used, right?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java [hunk 4]
```diff
@@ -37,6 +37,7 @@ public MvPlanContext optimize(MaterializedView mv,
         OptimizerConfig optimizerConfig = new OptimizerConfig(OptimizerConfig.OptimizerAlgorithm.RULE_BASED);
         optimizerConfig.disableRuleSet(RuleSetType.PARTITION_PRUNE);
         optimizerConfig.disableRuleSet(RuleSetType.SINGLE_TABLE_MV_REWRITE);
+        optimizerConfig.disableRuleSet(RuleSetType.ALL_MV_REWRITE);
         // INTERSECT_REWRITE is used for INTERSECT related plan optimize, which can not be SPJG;
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java:41]: Is this repeated?
```
optimizerConfig.disableRuleSet(RuleSetType.SINGLE_TABLE_MV_REWRITE);
optimizerConfig.disableRuleSet(RuleSetType.ALL_MV_REWRITE);
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/SingleTableMvRewriteRule.java (hunk 3 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/SingleTableMvRewriteRule.java:34]: its main logic is moved into BestMvSelector. So remove it here

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java (hunk 2 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:1451]: I think we can do it in next big version.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/pattern/Pattern.java (hunk 1 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/pattern/Pattern.java:146]: rule based pattern do not support multi join pattern, so add it here.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java (hunk 4 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java:40]: fixed

**Review by LiShuMing — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Pull request merged**